File: samples/opticalflow/python/opticalflow.py
import cv2
import numpy as np

width = 640
height = 480
cap = cv2.VideoCapture()
cap.set(cv2.CAP_PROP_FRAME_WIDTH, width)
cap.set(cv2.CAP_PROP_FRAME_HEIGHT, height)
cap.open(0)
prevgray = None
while True:
    ret, frame = cap.read()
    if not ret:
        break

    # gray scale
    gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
    if prevgray is None:
        prevgray = gray.copy()
        continue

    # calculate dense optical flow
    flow = cv2.calcOpticalFlowFarneback(prevgray, gray, None,
                                        0.5, 3, 15, 3, 5, 1.1,
                                        cv2.OPTFLOW_FARNEBACK_GAUSSIAN)

    # draw direction of flow
    for y in range(0, frame.shape[0], 20):
        for x in range(0, frame.shape[1], 20):
            flowatxy = flow[y, x]
            cv2.line(frame, (x, y), (int(
                x+flowatxy[0]), int(y+flowatxy[1])), (255, 255, 255), 2, cv2.LINE_AA)
            cv2.line(frame, (x, y), (int(
                x+flowatxy[0]), int(y+flowatxy[1])), (192, 0, 0), 1, cv2.LINE_AA)
            cv2.circle(frame, (x, y), 1, (255, 255, 255), 2, cv2.LINE_AA)
            cv2.circle(frame, (x, y), 1, (0, 133, 255), 1, cv2.LINE_AA)

    cv2.imshow("Dense Optical Flow", frame)

    key = cv2.waitKey(10)
    if key == 27:
        break

    prevgray = gray.copy()


cap.release()
cv2.destroyAllWindows()

File: samples/opticalflow/java/OpticalFlow.java
import org.bytedeco.javacpp.indexer.FloatIndexer;
import org.bytedeco.opencv.global.opencv_imgproc;
import org.bytedeco.opencv.global.opencv_video;
import org.bytedeco.opencv.global.opencv_videoio;
import org.bytedeco.opencv.opencv_core.Mat;
import org.bytedeco.opencv.opencv_core.Point;
import org.bytedeco.opencv.opencv_core.Scalar;
import org.bytedeco.opencv.opencv_videoio.VideoCapture;

public class OpticalFlow {
    public static void main(String[] args) {
        int width = 640;
        int height = 480;
        VideoCapture cap = new VideoCapture();
        cap.set(opencv_videoio.CAP_PROP_FRAME_WIDTH, width);
        cap.set(opencv_videoio.CAP_PROP_FRAME_HEIGHT, height);
        cap.open(0);
        Mat prevgray = null;
        while (true) {
            Mat frame = new Mat();
            boolean ret = cap.read(frame);
            if (!ret) break;

            // gray scale
            Mat gray = new Mat();
            opencv_imgproc.cvtColor(frame, gray, opencv_imgproc.COLOR_BGR2GRAY);

            if (prevgray == null || prevgray.empty()) {
                prevgray = gray.clone();
                releaseMat(frame, gray);
                continue;
            }

            // calculate dense optical flow
            Mat flow = new Mat();
            opencv_video.calcOpticalFlowFarneback(prevgray, gray, flow,
                    0.5, 3, 15, 3, 5, 1.1, opencv_video.OPTFLOW_FARNEBACK_GAUSSIAN);

            // draw direction of flow
            FloatIndexer flowatxy = flow.createIndexer();
            for (int j = 0; j < frame.rows(); j += 20) {
                for (int i = 0; i < frame.cols(); i += 20) {
                    float fi = flowatxy.get(j, i, 0);
                    float fj = flowatxy.get(j, i, 1);
                    opencv_imgproc.line(frame, new Point(i, j),
                            new Point((int) (i + fi), (int) (j + fj)),
                            Scalar.all(255), 2, opencv_imgproc.LINE_AA, 0);
                    opencv_imgproc.line(frame, new Point(i, j),
                            new Point((int) (i + fi), (int) (j + fj)),
                            new Scalar(192, 0, 0, 0), 1, opencv_imgproc.LINE_AA, 0);
                    opencv_imgproc.circle(frame, new Point(i, j), 1,
                            Scalar.all(255), 2, opencv_imgproc.LINE_AA, 0);
                    opencv_imgproc.circle(frame, new Point(i, j), 1,
                            new Scalar(0, 133, 255, 0), 1, opencv_imgproc.LINE_AA, 0);
                }
            }

            OpenCVFX.imshow("Dense Optical Flow", frame);
            if (OpenCVFX.waitKey() == 27) {
                break;
            }

            prevgray = gray.clone();

            // release mat
            releaseMat(frame, gray, flow);
        }
        cap.release();
        OpenCVFX.destroyAllWindows();
    }

    private static void releaseMat(Mat... mats) {
        for (Mat mat : mats) {
            if (mat != null) {
                mat.release();
                mat.close();
            }
        }
    }
}
